design/StableRegions.md

```diff
@@ -0,0 +1,358 @@
+ - While being heap-allocated, the object is also `stable` (can be stored in a `stable var`, etc).
+ - `RegionObject { id: u16, padding: u16, page_count: u32; vec_pages: Value }`
+ - Field `id` gives the Region's numerical id as an index into the `region` table.
+ - Padding is needed for alignment of larger fields.
+ - Field `page_count` gives the number of pages allocated to the Region.
+ - Field `vec_pages` points at a heap-allocated `Blob` value, and it works with `page_count`
+   to represent a growable vector that we call the region's **"access
+   vector"** (because "blocks vector" sounds a bit strange, and its
+   used to support O(1) access operations):
+   - the access vector has `vec_capacity` slots.
+   - each slot is a `u16`.
+   - the first `size_in_pages + 127 / 128` slots contain a valid page block ID for the region.
+   - during an upgrade, the access vectors get serialized and deserialized as data `Blobs` (as if no pointers are inside each).
+
+
+### region-blocks relation
+
+The `region-blocks` relation is not materialized into a table in stable memory (doing so with a pre-allocated table would be prohibitively large).
+
+Instead, this relation is represented in two ways at the same time:
+ 1. by the set of heap-allocated region objects, and their access vectors.  The access vectors provide O(1) store and load support.
+ 2. by the `block-region` table and the `region` table, which together are sufficient to recreate all of the heap-allocated region objects.
+
+In ordinary operation, the second feature is not required.  In the event of an upgrade failure, however, it could be vital (See `rebuild`).
+
+### block-region table
+
+ - purpose:
+   - relate a block ID ("page block ID") to its region (if any) and its position in that region (see `rebuild`).
+
+   - NB: The organization of this table is not useful for efficient
+     access calculations of load or store (they require a linear
+     search that would be prohibitively slow).  OTOH, this
+     table is suitable to do a "batch" rebuild of the dynamic heap-allocated vectors
+     in that table, if need be (see `rebuild`).
+
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BlockRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding block ID.
+
+### region table
+
+ - purpose:
+   - relate a region ID to its size in pages.
+
+   - NB: This table exists as a "stable backup" of the per-region fields
+     that exist in the set of region objects, except for the access vectors' elements (see `block-region` table for their data).
+
+ - 32768 entries (statically sized).
+ - 8 bytes per entry.
+ - entry type = `RegionBlocks { size_in_pages: Nat64; }`
+ - the location of each entry in the table gives its corresponding region ID.
+ - the `size_in_pages` field measures the size of the region in terms of _pages_, not blocks.
+ - to do: track bit per region of whether in-use or reclaimed-and-available, somehow.
```

Comment:
I am thinking a little bit on the implications when region GC would be introduced. Probably, it could be worth to already have a specific GC design in mind, to analyze whether the prepared GC information is sufficient and practical.

I see different possibilities (just some thoughts):
* Reserving track bits for the stable regions (as planned): The GC would rely on some initial state of these bits where at least all allocated regions are identified as such. A particular aspect to consider is what is needed when region garbage collection is introduced as a new feature. If the track bits are not yet prepared in the expected state, introducing the GC would require a stable memory version change and a special migration step to initialize these bits for the first time. Therefore, it could be beneficial to already (conservatively) set the track bits even before the GC is supported.

* Not reserving GC track bits in stable memory. The GC could construct the free list of stable regions in the heap after its first run by looking at which region objects are alive. A region is garbage if its region object has become garbage, or differently to classical objects, it lacks a corresponding region object in the heap. The latter occurs when an object becomes unreachable before an upgrade without that the GC had a chance to detect this before an upgrade. Considering this scenario would also include the case that a region has become garbage before the introduction of the region GC functionality.

Reply:
I was hoping we could have one 1 mark bit per region, stored in stable memory.
On GC, mark bits are cleared.
During mark phase, mark bits are set as Region objects are marked (and marked on grow)
After GC, all blocks whose region ids are unmarked are recorded as unowned by any block (or maybe added to a some designated FreeList Region). 
During a region grow the table or freelist is scanned for the first unowned block(s).
But maybe that's too naive.